### PR TITLE
III-5856 Add location name to fetch geocoords

### PR DESCRIPTION
**Pull request by grubolsch**
### Added
Added location name to fetching geocoords

Related to PR https://github.com/cultuurnet/appconfig/pull/518

---
Ticket: https://jira.uitdatabank.be/browse/III-5856

### CODE REVIEW

File: app/Event/EventGeoCoordinatesServiceProvider.php [hunk 1]
```diff
@@ -36,6 +36,8 @@ function () use ($container): GeoCoordinatesCommandHandler {
                     new FullAddressFormatter(),
                     new LocalityAddressFormatter(),
                     $container->get(GeocodingService::class),
+                    $container->get('event_jsonld_repository'),
+                    $container->get('config')['add_location_name_to_coordinates_lookup'] ?? false
```
Comment (LucWollants) [inline app/Event/EventGeoCoordinatesServiceProvider.php:40]: Don't forget to create a pull request on https://github.com/cultuurnet/appconfig and set `add_location_name_to_coordinates_lookup` to true for ACC, but still to false for TEST and PROD

File: app/Geocoding/GeocodingServiceFactory.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use Geocoder\Provider\GoogleMaps\GoogleMaps;
+use Geocoder\StatefulGeocoder;
+use Http\Adapter\Guzzle7\Client;
+use Psr\Log\LoggerInterface;
+
+class GeocodingServiceFactory
+{
+    private bool $addLocationNameToCoordinatesLookup;
+
+    public function __construct(bool $isNewFeature)
```
Comment (LucWollants) [inline app/Geocoding/GeocodingServiceFactory.php:16]: Maybe make the name of this variable more explicit and also use `$addLocationNameToCoordinatesLookup`

File: app/Geocoding/GeocodingServiceProvider.php [hunk 3]
```diff
@@ -24,25 +21,24 @@ public function register(): void
     {
         $container = $this->getContainer();
 
+        $geocodingService = new GeocodingServiceFactory(
```
Comment (LucWollants) [inline app/Geocoding/GeocodingServiceProvider.php:24]: This variable name is rather confusing, it should be `$geocodingServiceFactory`

File: src/Geocoding/EnrichedGeocodingService.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+namespace CultuurNet\UDB3\Geocoding;
+
+/**
+ * Like the DefaultGeocodingService, but we are also sending the location name
+ */
+class EnrichedGeocodingService extends AbstractGeocodingService
+{
+    protected function getKey(string $address, string $locationName): string
+    {
+        return trim($address . ' ' . $locationName);
```
Comment (LucWollants) [inline src/Geocoding/EnrichedGeocodingService.php:14]: Does it make more sense to start the key with the location name?

File: src/Geocoding/CachedGeocodingService.php [hunk 5]
```diff
@@ -24,9 +24,11 @@ public function __construct(GeocodingService $geocodingService, Cache $cache)
         $this->cache = $cache;
     }
 
-    public function getCoordinates(string $address): ?Coordinates
+    public function getCoordinates(string $address, string $locationName=''): ?Coordinates
     {
-        $encodedCacheData = $this->cache->fetch($address);
+        $key = empty($locationName) ? $address : trim($address . $address);
```
Comment (LucWollants) [inline src/Geocoding/CachedGeocodingService.php:29]: This feels like a copy/paste error `trim($address . $address);`

Comment (LucWollants) [reply src/Geocoding/CachedGeocodingService.php:29]: Also can the `getKey` not be used from the injected `geocodingService`?

**Comment by LucWollants** [comment]
I wonder if `UpdateGeoCoordinatesFromAddressCommandHandler` also needs to be updated with the location. Maybe something to discuss with @corneelwille .

File: src/Geocoding/CachedGeocodingService.php (hunk 5 above)
Comment (grubolsch) [reply src/Geocoding/CachedGeocodingService.php:29]: I will also add a unit test for testing the key

File: app/Place/PlaceGeoCoordinatesServiceProvider.php [hunk 6]
```diff
@@ -34,7 +34,9 @@ function () use ($container) {
                     $container->get('place_repository'),
                     new FullAddressFormatter(),
                     new LocalityAddressFormatter(),
-                    $container->get(GeocodingService::class)
+                    $container->get(GeocodingService::class),
+                    $container->get('event_jsonld_repository'),
```
Comment (LucWollants) [inline app/Place/PlaceGeoCoordinatesServiceProvider.php:38]: This is the wrong repository, it should be the place repo

File: app/Event/EventGeoCoordinatesServiceProvider.php (hunk 1 above)
Comment (grubolsch) [reply app/Event/EventGeoCoordinatesServiceProvider.php:40]: See PR https://github.com/cultuurnet/appconfig/pull/518

File: app/Geocoding/GeocodingServiceFactory.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        $this->addLocationNameToCoordinatesLookup = $isNewFeature;
+    }
+
+    public function createService(LoggerInterface $logger, ?string $googleMapApiKey): GeocodingService
```
[the remark this answers is not in the document]
Comment (grubolsch) [reply app/Geocoding/GeocodingServiceFactory.php]: I removed the nullable, but I am not sure if my solution is the best.
I now just use as the default value an empty string, should I instead throw an exception if no google key api was provided? Not sure what is best here.

File: src/Geocoding/CachedGeocodingService.php (hunk 5 above)
Comment (grubolsch) [reply src/Geocoding/CachedGeocodingService.php]: I renamed the function, added a unit test, and also made it reuse the method searchTerm() now required for the GeocodingService interface. I did this because the logic how to make the key was duplicate code.

File: src/Geocoding/AbstractGeocodingService.php [hunk 8]
```diff
@@ -50,6 +50,4 @@ public function getCoordinates(string $address, string $locationName=''): ?Coord
             return null;
         }
     }
-
-    abstract protected function getKey(string $address, string $locationName): string;
```
Comment (LucWollants) [inline src/Geocoding/AbstractGeocodingService.php]: Why was this removed?

File: app/Geocoding/GeocodingServiceFactory.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (LucWollants) [reply app/Geocoding/GeocodingServiceFactory.php]: I would rely on the default PHP exception if the config is missing and just get the key from the config without any checks.

**Comment by grubolsch** [comment]
> I wonder if `UpdateGeoCoordinatesFromAddressCommandHandler` also needs to be updated with the location. Maybe something to discuss with @corneelwille .

Dat lijkt me iets te zijn ivm de organizer, die hebben toch geen adres? Ik weet niet goed de context van deze commandHandler, dus kan dat moeilijk inschatten.

**Comment by LucWollants** [comment]
> > I wonder if `UpdateGeoCoordinatesFromAddressCommandHandler` also needs to be updated with the location. Maybe something to discuss with @corneelwille .
> 
> Dat lijkt me iets te zijn ivm de organizer, die hebben toch geen adres? Ik weet niet goed de context van deze commandHandler, dus kan dat moeilijk inschatten.

They have an address.

**Review by LucWollants — APPROVED** [review]
